### PR TITLE
cql3: time_uuid_fcts: validate time UUID

### PR DESCRIPTION
**Pull request by bhalevy**
Throw an error in case we hit an invalid time UUID
rather than hitting an assert.

Ref #5552

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

### CODE REVIEW

**Comment by bhalevy** [comment]
In v2:
- added UUID.is_timestamp
- added asserts for create_time
- added unit tests for time uuid
- added validation for time UUID
- deleted outdated FIXME comment
- added time uuid validation tests
- abstract_function_selector: provide assignment_testable_source_context

Test: unit(dev), cql_cast_test:CQLCastTest.cast_timeuuid_test cql_additional_tests:TestCQL.timeuuid_test(dev)

File: cql3/functions/time_uuid_fcts.hh [hunk 1]
```diff
@@ -62,6 +62,18 @@ make_now_fct() {
     });
 }
 
+static int64_t get_valid_timestamp(const data_value& ts_obj) {
+    auto ts = value_cast<db_clock::time_point>(ts_obj);
+    int64_t ms = ts.time_since_epoch().count();
+    auto nanos_since = utils::UUID_gen::make_nanos_since(ms);
+    if (!utils::UUID_gen::is_valid_nanos_since(nanos_since)) {
+        if (!utils::UUID_gen::is_valid_nanos_since(nanos_since)) {
```
Comment (psarna) [inline cql3/functions/time_uuid_fcts.hh]: why is the same condition checked twice?

Comment (bhalevy) [reply cql3/functions/time_uuid_fcts.hh]: hmm, this was unintentional.  Will fix.

**Comment by psarna** [comment]
I also see that there are unresolved conflicts with /next, so please rebase as well.

**Comment by bhalevy** [comment]
In v3:
- deleted extraneous call to `is_valid_nanos_since`
- deleted inadvertent update to seastar submodule

**will rebase after next is promoted**

File: cql3/functions/time_uuid_fcts.hh (hunk 1 above)
Comment (bhalevy) [reply cql3/functions/time_uuid_fcts.hh]: Fixed in v3

**Comment by bhalevy** [comment]
Rebased onto a74a82d4d245d75311c19d0d8fba404d5d75937e and retested.

**Comment by bhalevy** [comment]
ping @psarna

**Comment by psarna** [comment]
Looks good, I hope that I didn't miss any cases when these asserts fire from user input :) I'll fetch these changes, retest, and queue.

**Comment by bhalevy** [comment]
Dequeued due to cql_query_test failure in debug mode.

**Comment by nyh** [comment]
Now you see why I don't like to mark the target of a pull request as "next" - your pull request got automatically closed as soon as it entered next, and now there is no way to reopen it, and you need to create a different one.

**Comment by bhalevy** [comment]
> Now you see why I don't like to mark the target of a pull request as "next" - your pull request got automatically closed as soon as it entered next, and now there is no way to reopen it, and you need to create a different one.

True, that's an unfortunate outcome. Similarly, issues that get fixed by the pr are closed when it's merged to next, rather than when next is promoted to master. At least the issues can be reopened, unlike the pr.

**Pull request merged**